### ISSUE TITLE
why isn't `build-pathway` a Makefile?

### ISSUE THREAD

**Issue by shriram** [bootstrapworld/curriculum#467]
A lot of the logic in `build-pathway` looks make-like. I wonder why this wasn't written as a Makefile along with custom functions/scripts for the individual functions. Isn't this file much harder to maintain because someone will have to first understand the make-like logic? (It's currently about 1750 LoC.)

**Comment by ds26gte** [comment]
I wanted to do it as a Makefile starting out, and it is still in the back of mind to see if this can still be done. There is a lot of sheer procedural work being done by the `build-pathway` script -- the timestamp-based fragments are only a tiny part of it. (Also Git largely makes timestamps irrelevant, as the integrity of files is now based exclusively on their content, regardless of their timestamp.) Converting the build script into a Makefile would not simplify much, as the majority of the bash script would still end up carried over verbatim, i.e., in the action parts of the Makefile rules. The unnatural separation of timestamp and bash actions would make it more, not less, complicated, with probably an _increase_ in the lines of code.

As large as the bash script is now, it is actually outsourcing a lot of work to Racket scripts. Using Racket for the entirety of the script would have been feasible too, however, for speed reasons, we moved the "easier" parts to bash. Indeed, earlier iterations of the build script were smaller, as it did a lot more via Racket. The migration of some of the Racket bits to bash did make things uglier and the script of course got longer, but the price was felt to be worth it.

**Comment by schanzer** [comment]
@shriram curious to know what you think of @ds26gte 's response here. I should also tell you that in the next 2-3 months a chunk (the workbook generation) of this build script is going to be outsourced to client-side JS, _and_ that it's likely a lot of it will also be simplified as we move to a more dynamic output format.

Do you still see a makefile as an unquestionable win, especially in light of Dorai's comments?

**Comment by schanzer** [comment]
@shriram just letting you know we haven't forgotten about this!

We spent some time discussing the issue today, and it boils down to this:

* Dorai has done some incredibly clever stuff to make sure that things build in parallel, and to save and compare checksums so that nothing gets rebuilt unless it needs to be. The phases of the build are carefully ordered, too, such that parallelization happens only where it can be done correctly. And those phases build lots of intermediate data, which is used in later phases.
* According to Dorai, `make` is just not general-purpose enough to replicate these optimizations.
* If we're willing to sacrifice this speed, it could be done in `make`.
* However, we _really_ need speed in order to write curriculum. If the build takes a long time, we wind up batching edits so that we get feedback on more than one at a time. Unfortunately, batching also means a lot more room for mistakes! 

This move to JS libraries helped somewhat -- all of the PDF generation, collation, and batching has been moved out of `bash`. However, the remaining stuff just can't be easily translated.

**Comment by shriram** [comment]
Would tools like parallel make help?

I'm still concerned about the GIGANTIC technical debt this has incurred. I
don't think anyone can ever maintain this, especially not in a way that
doesn't introduce a ton of bugs.

**Comment by ds26gte** [comment]
Branch `makeitso` will contain the eventual `Makefile` to replace `build-pathway` (which is 3000+ LoC now).

Architecture: The user-facing `Makefile` calls a short sequence of sub-makefiles in order (named `Makefile.phase1`, `Makefile.phase2`, etc). This is because rules are dynamically generated (using makefile functions entirely, not an external language), and the sources on some of the later rules can't be deduced until some previous rules have fired. But the sub-makefiles are all static (not dynamically generated) and their number will be a very small integer (maybe 4).

The user-facing `Makefile` is in the top directory. All the subfiles used by it are located in `lib/maker`.

**Comment by schanzer** [comment]
@ds26gte I'm seeing proglang `codap` and `none` lessons get skipped during asciidoc generation. 

For examples, look at `commutative-property` and `codap-bar-charts-codap`

**Comment by ds26gte** [comment]
Some `proglang.txt` files were missing a closing newline, which makes 1-line files look like 0-line files to `wc`. Rewrote avoiding `wc`. (So you can keep writing text files that don't end in newlines if that's what you want to do.)

**Comment by schanzer** [comment]
Just thought I'd chime in to say that the new build system has far surpassed all of my expectations. It suffers from none of the parallelized gremlins that the old one did, appears to use half the LOC, is far easier for me to read and understand, and runs more than twice as fast. 

It's already my daily driver for work.

@ds26gte I think the next step is to update `./build-pathway` as a convenience wrapper for make, to provide familar mappings for things like LINKCHECK and BOOK. After that, is there anything on your list before we merge?

**Comment by schanzer** [comment]
@shriram using `make` instead of `bash`

**Comment by shriram** [comment]
But build-pathway is still in bash?

**Comment by schanzer** [comment]
Yes, but it's unused -- For now, Dorai's work co-exists alongside the old build system on this branch. We could delete build-pathway and it wouldn't change a thing.

He and I discussed hollowing it out and turning it into a simple wrapper for `make`, both so that Rachel and Flannery wouldn't have to adapt to a new system and so that we could expose a more stable interface as the `make` system evolves over time. That's not yet done.

**Comment by shriram** [comment]
So which file is the new makefile?

**Comment by schanzer** [comment]
@ds26gte before squashing and merging, run your commit message by @shriram. #STRONGFEELINGS

**Comment by ds26gte** [comment]
Hi Shriram,

There is a Makefile in the top-level directory in the makeitso branch.  It
includes some (static) sub-Makefiles in lib/maker, where most of the new
helper files, some very small Bash and Lua scripts also reside. You are of
course welcome and encouraged to scan these files in lib/maker, but as a
user, the interface is just typing `make` at the top directory.

For convenience, some variables can be set during the make call and these
are described in
https://ds26gte.github.io/curriculum-build/#_using_the_makefile

If we're committed to going the Makefile route -- Emmanuel has been testing
it from all angles and he looks convinced --, *the old build-pathway can be
obsoleted*, or turned into a wrapper script for old time's sake as E said.
The README.adoc file has to be changed to describe the new make régime --
I'm still debugging edge cases so the doc is admittedly lagging.

--d

**Comment by schanzer** [comment]
I notice that `contracts-wescheme` doesn’t show up in the graph (which is good!), `but commutative-property-none` does. Shouldn’t we be pruning that graph for _only_ canonical lesson names? A lot of our clientside JS depends on this

**Comment by ds26gte** [comment]
`commutative-property-none` showing up was not really a bug, because only the proglang `pyret` was treated specially as deserving no-suffix status. I've extended this special treatment now to the proglang `none`.  

There is an implicit assumption here that a lesson with proglang `none` doesn't have any other proglang in its `proglang.txt`, or at least doesn't have `pyret` as another proglang. Expect glitches if you have a `proglang.txt` listing both `none` and some other proglang. (I haven't enforced this as an error: maybe I should.)

In effect, `none` now means "no proglang at all" and not, as before, "a bogus stopgap proglang which we're calling `none`". It would perhaps be cleaner to associate an empty proglang list with this situation, but we're committed to having proglang = Pyret when nothing is specified, and it would be too legacy-disruptive, not to mention too verbose, to have to start specifying Pyret explicitly, since the overwhelming bulk of our content is Pyret, and this is the kind of boilerplate spec that authors don't remember. I trust this (already existing) compromise is acceptable.

**Comment by schanzer** [comment]
Totally! Thanks

**Comment by ds26gte** [comment]
The `Makefile`-based implementation has now been merged into branch `fall2023` with the merge commit 377d1dd6f

Doc about the Makefile(s) is included in-repo at  https://github.com/bootstrapworld/curriculum/blob/fall2023/lib/maker/doc/makefile-impl.adoc
